Feature profiles endpoint

BacLuc opened the pull request.

BacLuc committed:
api: add /profiles endpoint

The /users endpoint is now exposed to all users and
exposes all properties which may be seen by all authenticated users.
Currently, this is only the displayName.
The /profiles endpoint is exposed to related users,
with all properties a related user may see.
Patch is only allowed for the /profiles endpoint, where
the user can edit the properties exposed to related users.
displayName is currently a read only property.

Use the strict equality operator for 'object === user',
because an exception was thrown for too deep recursion
with the == operator.

Issue: #2148

BacLuc committed:
frontend/campRoleMixin: remove hack for not always embedded users

Because before, the users could not be fetched directly.
Thus, we had to ensure that the camp was loaded with
fetching a single camp.

Issue: #2148

BacLuc commented:
Stellt sich noch die Frage, wieviel embedding wir vom Camp wieder entfernen wollen:
https://github.com/ecamp/ecamp3/blob/573f56945dfdee939673394c9879dd7e85d89cea/api/src/Entity/Camp.php#L50

BacLuc commented on api/src/Repository/UserRepository.php:
```diff
@@ -34,9 +34,30 @@ public function upgradePassword(UserInterface $user, string $newEncodedPassword)
+     * @throws \Doctrine\ORM\NonUniqueResultException
+     * @throws \Doctrine\ORM\NoResultException
+     */
+    public function areUsersRelatedViaCamp(User $user1, User $user2): bool {
```
Ich habe den check in der DB implementiert, weil es sein könnte dass mit den CampCollaboration::getCamp und den Camp::getCollaborations zuviele queries auf die DB gehen.

carlobeltrame commented on api/src/Entity/User.php:
```diff
@@ -194,6 +206,13 @@ public function __construct() {
         $this->collaborations = new ArrayCollection();
```
Mir ist noch aufgefallen dass `$plainPassword` in der `read` Group ist. Hat keine Auswirkungen, da `readable: false`, aber sollten wir vielleicht dennoch entfernen, ist sonst nur verwirrend.

BacLuc replied:
done

carlobeltrame commented on api/src/Repository/UserRepository.php:
```diff
@@ -34,9 +34,30 @@ public function upgradePassword(UserInterface $user, string $newEncodedPassword)
+        $queryBuilder->andWhere('relatedCampCollaborations.user = :user2');
+        $queryBuilder->setParameter('user1', $user1);
+        $queryBuilder->setParameter('user2', $user2);
+        $queryBuilder->setParameter('status_established', CampCollaboration::STATUS_ESTABLISHED);
```
Beim `CampRoleVoter` überprüfe ich zusätzlich noch, ob die Rolle eine von `CampCollaboration::VALID_ROLES` ist. Ist zwar momentan nicht möglich dass etwas anderes in der DB gespeichert ist. Denkst du das wäre sinnvoll oder unnötig?

BacLuc replied:
Momentan ist es für diesen Check nicht wirklich wichtig, in welcher Rolle sie releated sind.
Hier ist vor allem der status wichtig, da wir die zugangsbeschränkung auf die user/profile properties nicht auf rollen aufbauen.
Dass in der DB eine ungültige Rolle ist (oder dass ein Hacker es nicht schafft, sich eine gültige zu geben) halte ich für keinen use case.

carlobeltrame commented on api/src/Serializer/Normalizer/UserNormalizer.php:
```diff
@@ -0,0 +1,55 @@
+    }
+
+    public function supportsNormalization($data, string $format = null) {
+        return $this->decorated->supportsNormalization($data, $format);
```
Eigentlich wäre die Idee in API Platform dass du in diesem Fall hier nicht dekorierst, sondern "normal" Composition anwendest: https://api-platform.com/docs/core/serialization/#changing-the-serialization-context-on-a-per-item-basis

In `supportsNormalization` würde dann der `$data instanceof User` Test stehen, plus ein Rekursions-Schutz. Und dafür würde `normalize` wirklich nur noch laufen, wenn `$object` ein `User` ist. Ein Beispiel dafür ist der `InputFilterDenormalizer`.

Das würde eigentlich auf diesen Normalizer sowie den ContentTypeNormalizer zutreffen, ich nehme an von dem hast du kopiert. Bin aber selber unsicher ob diese Unterscheidung von Dekoration und Komposition wirklich wichtig ist.

BacLuc replied:
Da bin ich überfragt.
Wir haben momentan noch keinen Normalizer mit Composition.

Die Beispiele zeigen wie man 
eine normalization Group aufgrund der Rolle hinzufügt für alle Items -> Decorated
den context auf per item basis ändert -> Composition

Mein Fall wird da nicht abgedeckt, und eventuell könnte man beides mit beiden varianten implementieren...?

Mein Fall wäre eher das Beispiel https://api-platform.com/docs/core/serialization/#decorating-a-serializer-and-adding-extra-data
(Wobei dafür könnte man auch unterschiedliche normalization_contexts verwenden)
Aber ich möchte ja unbedingt einen berechneten Link einfügen.
(Sonst müssten wir die property definitionen kopieren, oder 2 Entities mit inheritance haben, oder anderes komisches Zeugs).

Ich würde sagen, solange wir nicht wissen, wann wir Composition und wann Dekoration verwenden sollen, würde ich es überall möglichst gleich machen.

BacLuc replied:
Hast du das eventuell so gemeint. (Unterschied zwischen generellen normalizern für alle Ressourcen, und solchen, die nur für eine Resource gelten):

![grafik](https://user-images.githubusercontent.com/1506818/142768229-c3ca8f49-85dc-4dd7-b873-6f38a84699d4.png)
http://www.plantuml.com/plantuml/uml/bPDVI_DG3CRl-od2tejNBzhT3v6DYJgG8NFU2k4qdGTFSqgQuH_nkzkMuTdOcRMzJ9xaTvuadTQAejlI9V-VLQjwCXvN3hcDPCQNO-EmhjFSuaQmJ18XeyWRHp2w8IP11rQf1FPIehCV9FK8FXDeFlPAeBu2Ny0jkaKZM_we9d0jleHPPU7UeHPDeYlgSCVoNSklP4-qegQGyhbtZenQptUMNuvAsY-545wmaXHe278oNjeEnwtdLipMeQoe824sT4NQCdWJVzbWp130HPRrUUrwnYi7a9BcSVjB2vdPAmgj1Qt6zjRDBAhU3UTVT32y1puuh20-gy-XGS3SinBh-hsYxhh2eaBYzsDPqfi-L6GikjCr0zSOfxTJxyNrEdwiIGwjjf4_UuoAXvaBm8YjG0X-eWVmbARnCPsKTzm-NPPTzluJoPGutvRk6m00

carlobeltrame replied:
Ja, genau um diese Unterscheidung geht es. Ich hatte diese Patterns bisher so in der [API Platform Dokumentation](https://api-platform.com/docs/core/serialization/#changing-the-serialization-context-on-a-per-item-basis) und [SymfonyCasts](https://symfonycasts.com/screencast/api-platform-security/normalizer-aware) gesehen, und darum versucht zu übernehmen, bin aber selber nicht sicher was überhaupt die Begründung für die Unterscheidung ist.

So oder so fände ich es leicht schöner, wenn das grosse `if ($object instanceof User) {` von der Methode `normalize` in die Methode `supports` verschoben werden könnte. Das hat überhaupt meinen Kommentar ausgelöst.

Nebenbei: Ich glaube ich habe das Pattern selber nicht korrekt angewendet: Der `InputFilterNormalizer` den ich für die InputFilter `Trim`, `CleanHTML` etc. geschrieben habe greift allgemein bei allen Entities ein, verwendet aber Composition statt Decoration. Und der ContentTypeNormalizer von usu verwendet aktuell Decoration statt Composition. Trotzdem funktionierts bisher nicht schlecht, was auch ein Argument dafür ist, überall nur eine Variante zu verwenden.

carlobeltrame replied:
Wir lassens mal so bis es das nächste Mal wieder auftritt

carlobeltrame commented on api/tests/Api/Profiles/ReadProfileTest.php:
```diff
@@ -0,0 +1,105 @@
+        ]);
+    }
+
+    public function testGetSingleProfileIsAllowedForRelatedUser() {
```
Dieser Test müsste technisch gesehen `testGetSingleProfileIsAllowedForSelf` heissen, oder? Könntest du noch einen Test machen der wirklich einen anderen User im gleichen Camp abfragt?

BacLuc replied:
das ist ein anderer user.
Für die default credentials wird user1manager verwendet mit dem usernamen "test-user".
Siehe https://github.com/ecamp/ecamp3/pull/2241/files/c82161671d692c66e4fb0ae7b5607b1c3342fadc#diff-54d77a6ffe0b4a4b357273fe1edb39fdaa33ba74571f85e5907d4bfb118de979R32

carlobeltrame replied:
Aha :bulb: alles klar, merci

carlobeltrame commented on api/tests/Api/Users/DeleteUserTest.php:
```diff
@@ -18,13 +18,23 @@ public function testDeleteUserIsDeniedForAnonymousUser() {
+    public function testDeleteUserIsDeniedForDifferentUser() {
+        $user2 = static::$fixtures['user4unrelated'];
+        static::createClientWithCredentials()->request('DELETE', '/users/'.$user2->getId());
+        $this->assertResponseStatusCodeSame(403);
```
War der Test vorher falsch? Warum kam vorher 404 und jetzt 403? An den Security Rules und EntityFilters hast du doch gar nichts geändert?

BacLuc replied:
Der Entity Filter ist jetzt ganz weg.
User dürfen alle sehen.
Dann schreitet aber die security rule ein.

carlobeltrame commented on api/tests/Api/Users/ReadUserTest.php:
```diff
@@ -36,13 +44,34 @@ public function testGetSingleUserIsAllowedForSelf() {
+                ],
+            ],
+        ]);
+    }
```
Wie wärs mit einem Test dass gewisse Properties des Users jetzt in diesem Endpoint nicht mehr verfügbar sind? Dafür könntest du vielleicht auch einfach diesen Test auf `assertJsonEquals` umstellen.

BacLuc replied:
Da hast du recht, ist gemacht bei einem test.

carlobeltrame commented on api/tests/Security/Voter/UserIsRelatedVoterTest.php:
```diff
@@ -0,0 +1,113 @@
+        $this->assertEquals(CampRoleVoter::ACCESS_GRANTED, $result);
+    }
+
+    public function testDeniesAccessWhenUsersAreRelated() {
```
```suggestion
    public function testGrantsAccessWhenUsersAreRelated() {
```

BacLuc replied:
Uups, danke

carlobeltrame reviewed (dismissed):
Cool! Habs noch nicht durchgeklickt. Ein paar (hoffentlich) kleinere Anmerkungen.

BacLuc committed:
api/User: remove plainPassword from read group

pmattmann commented on api/src/Entity/User.php:
```diff
@@ -37,9 +38,18 @@
+        'get' => ['security' => 'is_fully_authenticated()'],
+        self::PROFILE_GET => [
+            'method' => 'GET',
+            'path' => 'profiles/{id}',
```
Können wir hier der Route wieder .{_format} anhängen?
(Siehe self::ACTIVATE oben)

Dadurch kann der Endpoint auch am Browser angesehen werden  /profiles/{id}.jsonhal

BacLuc replied:
Kann ich machen, sobald https://github.com/ecamp/ecamp3/pull/2241#discussion_r750920774 geklärt ist

BacLuc replied:
Ist nicht mehr nötig

pmattmann commented on api/src/Entity/User.php:
```diff
@@ -37,9 +38,18 @@
+        ],
+        self::PROFILE_PATCH => [
+            'method' => 'PATCH',
+            'path' => 'profiles/{id}',
```
Können wir hier der Route wieder .{_format} anhängen?
(Siehe self::ACTIVATE oben)

Dadurch kann der Endpoint auch am Browser angesehen werden  /profiles/{id}.jsonhal

BacLuc replied:
Kann ich machen, sobald https://github.com/ecamp/ecamp3/pull/2241#discussion_r750920774 geklärt ist

BacLuc replied:
Ist nicht mehr nötig

usu commented on api/src/Serializer/Normalizer/UserNormalizer.php:
```diff
@@ -0,0 +1,55 @@
+            $profileUri = $this->uriTemplate->expand($profileTemplate, ['id' => $object->getId()]);
+            $userUri = $this->uriTemplate->expand($uriTemplate, ['id' => $object->getId()]);
+
+            $data['_links']['self']['href'] = $userUri;
```
Schön wäre, wenn wir zu Profile nicht nur den Link haben, sondern die Daten auch embedden können. Wenn wir eine CampCollaboration Liste von 20 Leitern haben setzten wir ansonsten 20 Einzelrequests ab, um die Details zu laden.

Wollen wir User und Profile nicht einfach als separate Entities modellieren? Dann kommt das halt alles gratis mit.

BacLuc replied:
Brauchen wir momentan (halt nur momentan) nicht, da wir auf Collaborator.vue nur den DisplayName vom User anzeigen. (Aber die Migration in der Produktion wäre möglich, wahrscheinlich schon nicht so toll).

Da ist dann wieder die Frage:

1. Seperater Endpoint für /profile, oder "subresource" (nur embedded im User falls berechtigung da)
2. Link/Embedding nur Einfügen, wenn die Berechtigung da ist?
2. Als abstrakte Klasse oder als 1:1 Relation?
3. Wenn 1:1 Relation, sowohl bei User als auch bei Profile den Link einbauen?
4. Für jede Berechtigungstufe eine eigene Klasse machen? Welche Klassen verlinken/embedden zu welchen?

carlobeltrame replied:
Ich sehe es glaube ich ähnlich wie usu, habe aber noch nicht alles 100% durchgedacht. Ich versuche mal, deine Fragen aus meiner Sicht zu beantworten.
1. Separate Entities (oder separate ApiResources mit gemeinsamer Entity), bedeuten für mich automatisch separate Endpoints
2. Links würde ich immer einfügen, da es im Frontend nicht trivial ist, herauszufinden ob ein Link wirklich nicht existiert (während loading oder reloading kann man das nicht wissen, und ein reload kann auch während einem delete einer anderen Entity getriggert werden, etc...). Embedding natürlich nur wenn die Berechtigung da ist.
3. Mein Vorschlag wäre eine gemeinsame Tabelle, gemeinsame Doctrine Entity, aber mehrere ApiResources die dieselbe Entity verwenden. Das ist möglich via Custom [DataProvider](https://api-platform.com/docs/core/data-providers/) (lesend) und [DataPersister](https://api-platform.com/docs/core/data-persisters/) (schreibend), aber hier bin ich eben noch nicht sicher ob das die einfachste Lösung ist (funktionieren dann z.B. ApiFilter out of the box?). Wenns mit Vererbung einfacher geht, why not. Eine Relation mit komplett separaten Tabellen wäre eher ein last resort für mich, wegen der zusätzlichen Komplexität im Datenmodell.
4. Am liebsten würde ich der Bequemlichkeit halber von jeder Variante auf jede Variante verlinken. Es könnte sein dass das dann ab 3 Varianten zu unendlich vielen JOINs führt in der EagerLoadingExtension. Falls das der Fall ist würde ich die "öffentliche" Variante als Einstiegspunkt nehmen und von dort auf alle anderen Varianten verlinken.
5. Dito 4.

manuelmeister replied:
## Core Meeting Decision

Wir gehen mit der Variante: "Separierte Entitäten mit 1:1 Relation"
Es gibt immer beide.
Wir posten auf den User und patchen auf den Profile Endpoint.

BacLuc replied:
done

BacLuc committed:
InputFilterDenormalizer: also filter 1:1 related entities in the body

That is needed that it's possible to create a profile when
creating a user.
Else the InputFilter on the profile properties are not applied.

Extract new functionality to separate method.
Rename previous function getInputFilterAttributes to createInputFilterAttributesFrom,
because it's a factory method to create the InputFilter instances.
Extract getFilterAttributes method (which now gets the filter attributes from the whole class).

Issue: #2148

BacLuc committed:
UpdateUserTest: fix 'Nesting level too deep - recursive dependency?'

Nesting level too deep - recursive dependency? in /srv/api/vendor/symfony/expression-language/Node/BinaryNode.php on line 122

BacLuc committed:
UpdateProfileTest: fix 'Nesting level too deep - recursive dependency?'

Nesting level too deep - recursive dependency? in /srv/api/vendor/symfony/expression-language/Node/BinaryNode.php on line 122

usu commented on api/src/Entity/Profile.php:
```diff
@@ -0,0 +1,137 @@
+    #[InputFilter\Trim]
+    #[InputFilter\CleanHTML]
+    #[ApiProperty(example: self::EXAMPLE_FIRSTNAME)]
+    #[Groups(['read', 'write'])]
```
I'd make this `#[Groups(['read', 'create'])]` only until someone implements #2357

BacLuc replied:
Habe ich jetzt einfach verschoben. Ich schlage vor wir lassen das so und implementieren #2357, das brauchen wir so oder so vor dem go live.

usu replied:
Nicht sicher, ob #2357 zwingend Teil des MVPs sein muss (man kommt wahrscheinlich gut ein paar Monate ohne Wechsel von E-Mailadresse aus bzw. die wenigen Fälle könne man im Notfall auch manuell handeln). Ich würde eher für den Moment ausschalten. Wenn jemand dann #2357 vor Go-Live implementiert, perfekt. Wenn nicht, ist dann auch nicht so schlimm.

BacLuc replied:
Ok, mache ich aber in einem separaten PR.

usu replied:
Ok for me

usu commented on api/src/Serializer/Denormalizer/InputFilterDenormalizer.php:
```diff
@@ -63,10 +64,27 @@ protected function filterInputs($data, string $className): array {
+                        $toFilterRelatedEntities[] = [$propertyName, $type];
+
+                        break;
+                    }
```
Was macht der Code oben genau? Docu/comments möglich?

BacLuc replied:
na ja.
Man könnte es in eine Methode auslagern, und dann
```
$filterAttributes[] = [$property->name, $filterAttribute];
foreach ($reflClass->getProperties() as $property) {
            if ($property->getDeclaringClass()->name === $className) {
                foreach ($this->getInputFilterAttributes($property) as $filterAttribute) {
                    $filterAttributes[] = [$propertyName, $filterAttribute];
                }
            }
            $tuple = getTupleForRelationProperty($property)
            if($tuple != null) {
                $toFilterRelatedEntities[] = $tuple;
             }
```

Oder sonst müssen die tests reichen..?

carlobeltrame replied:
Soweit ich sehe ist der neue Code dafür zuständig, auch die InputFilters in related Entities zu evaluieren.

Wie wäre es, wenn wir, anstatt das magisch für alle Relations zu machen, auch auf die `Assert\Valid` Annotation von Symfony achten würden, oder eine eigene Annotation `InputFilter\Relation` o.ä. dafür schaffen würden?

So oder so fände ich den Code glaube ich lesbarer, wenn du das Sammeln der zu evaluierenden Relations in einem separaten `foreach`-Loop machen würdest. Dieser separate Loop könnte dann auch gut in eine eigene, sauber benannte und dokumentierte Methode ausgelagert werden.

BacLuc replied:
Aus security sicht möchte ich eigentlich InputFilter und Validations nicht einschalten müssen.
Sondern falls jemand einen guten grund hat, das auszuschalten, kann er das. Aber per default werden Validations und Inputfilter aktiviert und überprüft.

Ich würde das eher noch ausbauen, dass es das automatisch für collections auch noch macht (mit der template Annotation/Attribute hoffentlich). Und dann das in 3 verschiedene Denormalizer packen (1 Denormalizer für eine Entity, 1 Denormalizer für OneToOne und ManyToOne, 1 denormalizer für ManyToOne und ManyToMany, und evtl später noch für Embeddable Entities, falls wir das mal brauchen.)

Das aufteilen auf 2 loops mache ich gerne noch.

usu replied:
> Soweit ich sehe ist der neue Code dafür zuständig, auch die InputFilters in related Entities zu evaluieren.

Ah, es geht um denormalizing von embedded entities (writableLink: true). Das war mir jetzt beim Überfliegen des Codes nicht klar. Ein paar Kommentarzeilen wären hilfreich.

Ansonsten mit Kommentar von @carlobeltrame good for me.

BacLuc replied:
Ist ausgelagert in eine separate methode

usu commented on api/tests/Api/Users/ListUsersTest.php:
```diff
@@ -21,7 +21,7 @@ public function testListUsersIsAllowedForLoggedInUserButFiltered() {
         $this->assertResponseStatusCodeSame(200);
         $this->assertJsonContains([
-            'totalItems' => 1,
+            'totalItems' => 8,
```
Ist die Frage, ob wir dennoch uneingeschränktes Auflisten aller User ohne Einschränkung mittels Suchparameter oder ähnlichem erlauben wollen. Wir geben jetzt keine wesentlichen Daten mehr raus, aber ist halt dennoch ein zusätzlicher Attack-Vektor, wenn man schnell alle User auflisten kann. Kann von mir aus aber gerne in eine separate Diskussion bzw. separates Issue.

BacLuc replied:
Wir könnten die collection GET Operation abschalten. Evtl wollen wir mal noch nach usern suchen, können es ja dann wieder einschalten. Wir brauchen momentan nur die item GET Operation um referenzen aufzulösen.

Meinungen?

carlobeltrame replied:
Ich glaube fürs MVP brauchen wir dann sowieso eine Suchfunktion, womit die GET Collection Operation wieder nötig wird. Daher würde ich sie jetzt nicht extra nochmals entfernen.
Ich verstehe die Frage von usu eher so, ob es möglich ist, das Abrufen aller User ohne jegliche Filter zu verbieten, aber das Abrufen mit Query Parametern zu erlauben. So könnte die Angriffsfläche für DoS-Attacken verkleinert werden.

BacLuc replied:
Für DoS eher weniger, dann würde man das bei allen Entities machen.
Wahrscheinlich eher um nicht alle User ID's zu leaken.

Ich denke fürs MVP brauchen wir keine Suchfunktion. Diese haben wir ausgebaut, damit wir einen einzigen Ablauf haben, wie ein neues Mitglied ins camp kommt (per E-Mail). Für das MVP denke ich reicht es völlg, wenn man die E-Mail der Leiter in das Input feld kopiert, und so die Leiter einlädt.

Daran denken, dass beim abschalten der Endpoint nicht mehr im HalEntryPointNormalizer drin ist. #2347

usu replied:
> Wahrscheinlich eher um nicht alle User ID's zu leaken.

Ja genau, um das geht es mir (UserIDs und auch alle DisplayNames).

> Ich denke fürs MVP brauchen wir keine Suchfunktion.

Ok. für mich. Ich würde in diesem Fall dafür voten, die GET Operation für den Moment auszuschalten.

BacLuc replied:
Ist disabled

usu reviewed (approved):
LGTM. Kleinere Kommentare siehe unten.

carlobeltrame commented on api/src/Entity/Profile.php:
```diff
@@ -0,0 +1,137 @@
+ */
+#[ApiResource(
+    collectionOperations: [
+        'get' => ['security' => 'false'],
```
Das ergibt doch einen 403, oder? 403 bedeutet typischerweise, dass die Operation mit der richtigen Authentifizierung erlaubt wäre. Könntest du die Operation ganz weglassen (`collectionOperations: []`) damit hier ein 405 Method not allowed kommt?

BacLuc replied:
Dann wird der Endpunkt im halentrypointnormalizer nicht angezeigt. #2347

carlobeltrame commented on api/src/Entity/User.php:
```diff
@@ -28,6 +28,7 @@
             'security' => 'true', // allow unauthenticated clients to create (register) users
             'input_formats' => ['jsonld', 'jsonapi', 'json'],
             'validation_groups' => ['Default', 'create'],
+            'normalization_context' => ['groups' => ['read', 'User:create']],
```
Sehe ich das richtig dass du hier `User:create` statt `create` schreiben musst, damit das dazugehörende Profile nicht z.B. beim Erstellen einer CampCollaboration exponiert wird?

BacLuc replied:
Yes

carlobeltrame commented on api/src/Entity/User.php:
```diff
@@ -236,18 +184,22 @@ public function eraseCredentials() {
 
     public function getUserIdentifier(): ?string {
-        return $this->username;
+        return $this->profile->username;
```
Weisst du ob es in Symfony / Doctrine nötig bzw. möglich ist, beim Laden von Usern immer das Profile auch eager mitzuladen? Aus dem Bauch heraus habe ich das Gefühl dass die Authenticator Logik nicht davon ausgeht dass der Identifier in einer anderen Entity liegt, und dass es sich lohnen könnte, die immer zusammen zu laden. In [Laravel](https://laravel.com/docs/8.x/eloquent-relationships#eager-loading-by-default) und [Rails](https://stackoverflow.com/a/5072406) kenne ich extra Möglichkeiten dazu, um n+1-Queries zu vermeiden, daher frage ich.

BacLuc replied:
Die Authenticator Logik geht nicht davon aus, deshalb wird auch die custom query in (
https://github.com/ecamp/ecamp3/pull/2241/files#diff-becf2a543e0d61512331e058d4a261eb51dde403d296b45eac7c2bda1c03e23b
https://github.com/ecamp/ecamp3/pull/2241/files#diff-0b98959784a71cdf8e8285f245c68df43aca9c763a8a80a240c1bead544a996b , Zeile 42 in der neuen version)
 verwendet.
Es gibt nicht nur eine unschöne n+1 query, sondern funktioniert gar nicht, weil die implementation vom authentication bundle nicht über das UserInterface geht, sondern direkt die property auf der Entity User queried.

Getestet habe ich es nicht, aber anscheinend lädt doctrine die inverse side eagerly, wenn man die owning side lädt. (https://stackoverflow.com/questions/12362901/doctrine2-one-to-one-relation-auto-loads-on-query)
In diesem PR ist die owning side der User (https://github.com/ecamp/ecamp3/pull/2241/files#diff-a7ca47174e57c7008ac18ddf6023fbbfd8c5beb8ddb107a005405a87e291bce7 , Zeile 118).
Dies ist umgekehrt wie im beispiel von stackoverflow.
https://www.doctrine-project.org/projects/doctrine-orm/en/2.9/reference/association-mapping.html#one-to-one-bidirectional

carlobeltrame commented on api/src/Entity/User.php:
```diff
@@ -70,74 +71,6 @@ class User extends BaseEntity implements UserInterface, PasswordAuthenticatedUse
     #[ApiProperty(readable: false, writable: false)]
     public Collection $collaborations;
```
`ownedCamps`, `password` und weitere Properties hängen jetzt immer noch am User, sind also nur daher nicht für die "Öffentlichkeit" exponiert weil sie readable: false sind. Ist das korrekt so, oder wären die nicht besser beim Profile aufgehoben?

Denselben Gedanken hatte ich für die `collaborations`, aber die müssen ja auf dem User sein damit man neue User in ein Lager einfügen kann. Oder wird das durch das Einladungs-System abgefangen, und auch die könnte man aufs Profile verschieben?

BacLuc replied:
Wenn ich als zukünftiger Entwickler die Applikation anschaue, würde ich weder username noch passwort auf der Profile Entity erwarten. Dies ist jetzt nur so komisch, weil wir diese Abstufung der Sichtbarkeit wollen.

Ich habe mir sogar überlegt, den username auf dem User zu lassen, und dann vom Profile her über den User zu holen, habe mich dann aber dagegen entschieden, da das noch komischer wäre.

Ownedcamps denke ich können wir bald auch abschaffen, das braucht es nicht wirklich momentan. Würde ich aber nicht in diesem PR machen.

Jetzt habe ich einfach das verschoben, was wir nur auf dem profile wollen. Alles was so oder so nicht sichbar ist (und nicht sichtbar bleiben wird), bleibt beim User..
Das verschieben wäre jetzt nicht mehr so schwierig, da kann man die migration analog schreiben.
Die Idee momenan war (leider nicht ganz durchgezogen), dass aus sicht des php codes der User noch so funktioniert wie bisher. Das macht es auch einfacher, wenn wir das Profile dann noch weiter aufteilen wollen.

Die collaborations aufs profile zu schieben, weil wir die sichtbarkeit der properties abstufen wollen, macht für mich nicht wirklich sinn. Die beziehung besteht zum user, und nicht zum profile. Das auflisten der "campCollaborations" haben wir schon, indem wir die für den eigenen user sichtbaren camps filtern. Und dass man von einem anderen User sehen möchte, in welchen camps der ist, wird so bald nicht nötig sein, denke ich.

"
Denselben Gedanken hatte ich für die collaborations, aber die müssen ja auf dem User sein damit man neue User in ein Lager einfügen kann. Oder wird das durch das Einladungs-System abgefangen, und auch die könnte man aufs Profile verschieben?
"
Könnte man, das Einladungssystem fängt das nicht ab, aber man könnte das so schreiben. Macht für mich aber nicht mehr sinn, als so wie es momentan ist.

BacLuc committed:
InputFilterDenormalizer: make private methods private

BacLuc committed:
api/User: disable collection get operation

We don't need it and it would expose all user ids.
